### PR TITLE
feat: improved guided tour

### LINKED ISSUE #9636: [Bug]: When the widget is dragged & dropped manually in Welcome tour the widget shows error
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When a user follows the _Welcome tour_ and does the tour manually by following the instructions of the tour when the table widget is dropped on the canvas it shows error and continues to be present even after the right binding (Refer video)

image

image

### Steps To Reproduce

1. In Homepage click on Welcome tour present at the bottom left
2. Follow instructions > _Let's go_ > _Start Building_ > _New Query_ > _Run_ > _Add Widget_ > Drag & Drop table widget and observe

### Environment

Production

### Version

Cloud

### LINKED ISSUE #9639: [Bug]: Not able to proceed to next step after user binds the data from suggested widgets in Welcome tour
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When a user follows the _Welcome tour_ and does the tour manually by following the instructions of the tour when the user selects any widget that appears below 'Suggested widgets' on the right and selects the Table widget, the user is not able to continue the tour post binding the data. 'Do it for me' performs no action. (Refer video)

https://loom.com/share/8350bc286eb949a2ab6561bceaf144bd

### Steps To Reproduce

1. In Homepage click on Welcome tour present at the bottom left
2. Follow instructions > Let's go > Start Building > New Query > Run > Under _Suggested Widgets_  click on Table Widget > Observe the table and bindings > Click on _Do it for me_ (no action is seen)

### Environment

Production

### Version

Cloud

### LINKED ISSUE #9641: [Bug]: When the user uses the 'Suggested widgets' the bindings fails for Select widget in Welcome tour
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When a user follows the _Welcome tour_ and does the tour manually by following the instructions of the tour when the user selects the _Select_ widget that appears below 'Suggested widgets' on the right, the binding fails and no data is shown. (Refer video)

https://loom.com/share/5948225656234aa5be50984d109cfbed

### Steps To Reproduce

1. In Homepage click on Welcome tour present at the bottom left
2. Follow instructions > Let's go > Start Building > New Query > Run > Under Suggested Widgets click on _Select_ Widget > Observe the widget drop down and bindings >

### Environment

Production

### Version

Cloud

### LINKED ISSUE #10172: DuplicateKeyException: E11000 duplicate key error collection: appsmith.application index: organization_application_delet...
Onvoarding flow for new user is throwing duplicate key exception for `Super Standup app`. More details can be found out in sentry logs

Sentry Issue: [APPSMITH-1GJ](https://sentry.io/organizations/appsmith/issues/2887535067/?referrer=github_integration)

```
MongoWriteException: E11000 duplicate key error collection: appsmith.application index: organization_application_deleted_gitApplicationMetadata_compound_index dup key: { organizationId: "61d545cba4437918c021e3bb", name: "Super Standup", deletedAt: null, gitApplicationMetadata.remoteUrl: null, gitApplicationMetadata.branchName: null }
    at com.mongodb.reactivestreams.client.internal.MongoOperationPublisher.lambda$createSingleWriteRequestMono$24(MongoOperationPublisher.java:491)
    at reactor.core.publisher.Mono.lambda$onErrorMap$30(Mono.java:3680)
    at reactor.core.publisher.Mono.lambda$onErrorResume$32(Mono.java:3770)
    at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:94)
    at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:192)
...
(63 additional frame(s) were not displayed)

DuplicateKeyException: E11000 duplicate key error collection: appsmith.application index: organization_application_deleted_gitApplicationMetadata_compound_index dup key: { organizationId: "61d545cba4437918c021e3bb", name: "Super Standup", deletedAt: null, gitApplicationMetadata.remoteUrl: null, gitApplicationMetadata.branchName: null }; nested exception is com.mongodb.MongoWriteException: E11000 duplicate key error collection: appsmith.application index: organization_application_deleted_gitApplicationMetadata_compound_index dup key: { organizationId: "61d545cba4437918c021e3bb", name: "Super Standup", deletedAt: null, gitApplicationMetadata.remoteUrl: null, gitApplicationMetadata.branchName: null }
    at org.springframework.data.mongodb.core.MongoExceptionTranslator.translateExceptionIfPossible(MongoExceptionTranslator.java:106)
```

### PR DESCRIPTION
**Pull request by akash-codemonk**
Related to #7758 

Fixes #9641 
Fixes #9639
Fixes #9636
Fixes #10172 

## Description

Revamped welcome tour.

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Cypress test

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feat/onboarding-v2 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.05 **(-0.04)** | 36.59 **(-0.29)** | 34.95 **(-0.06)** | 55.46 **(-0.11)**
 :red_circle: | app/client/src/actions/onboardingActions.ts | 68.97 **(-1.3)** | 100 **(0)** | 5.26 **(-3.07)** | 53.85 **(-2.15)**
 :red_circle: | app/client/src/components/ads/Icon.tsx | 51.65 **(-0.67)** | 34.3 **(-1.03)** | 100 **(0)** | 51.54 **(-0.67)**
 :red_circle: | app/client/src/components/editorComponents/ActionNameEditor.tsx | 24.29 **(-0.71)** | 3.77 **(0.13)** | 0 **(0)** | 25.76 **(-0.8)**
 :red_circle: | app/client/src/components/editorComponents/ActionCreator/index.tsx | 39.11 **(-1)** | 6.96 **(0.24)** | 40 **(0)** | 38.98 **(-1.02)**
 :green_circle: | app/client/src/constants/DefaultTheme.tsx | 96.55 **(0.03)** | 90 **(0)** | 85.19 **(0)** | 96.04 **(0.04)**
 :green_circle: | app/client/src/constants/messages.ts | 74.94 **(0.93)** | 100 **(0)** | 25.75 **(2.69)** | 80.22 **(1.09)**
 :red_circle: | app/client/src/pages/Editor/EditorHeader.tsx | 86.23 **(-0.29)** | 50.68 **(0.01)** | 62.5 **(0)** | 85.71 **(-0.32)**
 :red_circle: | app/client/src/pages/Editor/PropertyPaneTitle.tsx | 70.21 **(-8.74)** | 21.05 **(-2.48)** | 71.43 **(0)** | 68.18 **(-8.96)**
 :green_circle: | app/client/src/pages/Editor/ToggleModeButton.tsx | 38.3 **(0.17)** | 12.35 **(-0.31)** | 17.24 **(0)** | 37.69 **(0.19)**
 :red_circle: | app/client/src/pages/Editor/WidgetSidebar.tsx | 70 **(-5.51)** | 30 **(-35.22)** | 44.44 **(0)** | 68.42 **(-6.05)**
 :red_circle: | app/client/src/pages/Editor/DataSourceEditor/Connected.tsx | 60.87 **(-3.13)** | 20 **(0)** | 0 **(0)** | 60.87 **(-3.13)**
 :red_circle: | app/client/src/pages/Editor/DataSourceEditor/DBForm.tsx | 53.33 **(-1.51)** | 22.22 **(0)** | 9.09 **(0)** | 58.49 **(-1.51)**
 :red_circle: | app/client/src/pages/Editor/DataSourceEditor/FormTitle.tsx | 25.58 **(-2.08)** | 5.41 **(0.28)** | 0 **(0)** | 26.83 **(-2.06)**
 :green_circle: | app/client/src/pages/Editor/DataSourceEditor/NewActionButton.tsx | 40 **(1.54)** | 4.76 **(0.21)** | 0 **(0)** | 45.16 **(3.49)**
 :green_circle: | app/client/src/pages/Editor/Explorer/index.tsx | 100 **(5.26)** | 100 **(33.33)** | 100 **(0)** | 100 **(5.26)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Actions/ActionEntityContextMenu.tsx | 39.13 **(-4.11)** | 0 **(0)** | 0 **(0)** | 39.13 **(-4.11)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Actions/MoreActionsMenu.tsx | 33.33 **(0)** | 25 **(-8.33)** | 0 **(0)** | 35.71 **(-0.4)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Datasources/DatasourceEntity.tsx | 45.95 **(8.45)** | 0 **(0)** | 0 **(0)** | 45.95 **(8.45)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Datasources/DatasourceStructureContainer.tsx | 46.67 **(-6.27)** | 0 **(0)** | 0 **(0)** | 46.67 **(-6.27)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Entity/index.tsx | 90.12 **(-3.03)** | 76.79 **(-2.84)** | 83.33 **(0)** | 89.74 **(-3.12)**
 :red_circle: | app/client/src/pages/Editor/Explorer/Widgets/WidgetContextMenu.tsx | 72.73 **(-3.59)** | 14.29 **(-2.38)** | 66.67 **(0)** | 72.09 **(-3.59)**
 :green_circle: | app/client/src/pages/Editor/Explorer/Widgets/WidgetGroup.tsx | 93.55 **(0.45)** | 70.83 **(0.62)** | 66.67 **(0)** | 93.33 **(0.47)**
 :red_circle: | app/client/src/pages/Editor/FirstTimeUserOnboarding/IntroductionModal.tsx | 93.62 **(0.29)** | 100 **(0)** | 75 **(-5)** | 93.02 **(0.52)**
 :green_circle: | app/client/src/pages/Editor/FirstTimeUserOnboarding/Statusbar.tsx | 90.11 **(0.75)** | 77.46 **(0)** | 85.71 **(0)** | 89.89 **(0.76)**
 :sparkles: :new: | **app/client/src/pages/Editor/GuidedTour/Boxed.tsx** | **78.57** | **25** | **100** | **78.57**
 :sparkles: :new: | **app/client/src/pages/Editor/GuidedTour/EndTour.tsx** | **52.94** | **100** | **0** | **52.94**
 :sparkles: :new: | **app/client/src/pages/Editor/GuidedTour/Guide.tsx** | **23.39** | **19.26** | **0** | **25.22**
 :sparkles: :new: | **app/client/src/pages/Editor/GuidedTour/constants.tsx** | **43.1** | **100** | **4.76** | **43.1**
 :sparkles: :new: | **app/client/src/pages/Editor/GuidedTour/useComputeCurrentStep.ts** | **6.82** | **0** | **0** | **6.82**
 :sparkles: :new: | **app/client/src/pages/Editor/GuidedTour/utils.ts** | **15.56** | **0** | **13.33** | **16.09**
 :red_circle: | app/client/src/pages/Editor/IntegrationEditor/DatasourceCard.tsx | 56.58 **(-1.11)** | 53.57 **(0)** | 0 **(0)** | 56.58 **(-1.11)**
 :green_circle: | app/client/src/pages/Editor/PropertyPane/Generator.tsx | 90.48 **(1.01)** | 78.57 **(8.57)** | 100 **(0)** | 94.74 **(0.62)**
 :red_circle: | app/client/src/pages/Editor/PropertyPane/PropertyControl.tsx | 51.91 **(-1.08)** | 34.88 **(-1.48)** | 21.43 **(0)** | 52.85 **(-1.12)**
 :green_circle: | app/client/src/pages/Editor/PropertyPane/PropertyPaneConnections.tsx | 67.01 **(0.34)** | 31.82 **(0.34)** | 48.48 **(0)** | 67.37 **(1.07)**
 :red_circle: | app/client/src/pages/Editor/QueryEditor/EditorJSONtoForm.tsx | 34.96 **(0.14)** | 27.47 **(-0.78)** | 0 **(0)** | 36.24 **(0.13)**
 :red_circle: | app/client/src/pages/Editor/WidgetsEditor/index.tsx | 93.85 **(0.2)** | 62.07 **(-0.89)** | 100 **(0)** | 93.55 **(0.22)**
 :red_circle: | app/client/src/reducers/uiReducers/onBoardingReducer.ts | 26.09 **(-0.83)** | 0 **(-100)** | 6.25 **(1.25)** | 26.09 **(-0.83)**
 :red_circle: | app/client/src/sagas/ApiPaneSagas.ts | 15.6 **(-0.47)** | 1.82 **(0.03)** | 7.14 **(0)** | 17.65 **(-0.58)**
 :red_circle: | app/client/src/sagas/ApplicationSagas.tsx | 14.92 **(-0.43)** | 1.01 **(-0.05)** | 4 **(0)** | 17.96 **(-0.36)**
 :green_circle: | app/client/src/sagas/DatasourcesSagas.ts | 13.39 **(0.1)** | 2.36 **(-0.04)** | 8.7 **(0)** | 15.94 **(0.19)**
 :green_circle: | app/client/src/sagas/OnboardingSagas.ts | 21.9 **(5.44)** | 1.23 **(-0.09)** | 4.35 **(-6.56)** | 24.39 **(7.67)**
 :green_circle: | app/client/src/sagas/PageSagas.tsx | 24.2 **(0.32)** | 5.88 **(-0.22)** | 14.71 **(0)** | 25.94 **(0.16)**
 :green_circle: | app/client/src/sagas/WidgetAdditionSagas.ts | 69.57 **(0)** | 61.22 **(1.22)** | 62.5 **(0)** | 70.49 **(0)**
 :green_circle: | app/client/src/sagas/WidgetDeletionSagas.ts | 42.77 **(2.1)** | 26.39 **(1.39)** | 42.11 **(0)** | 43.57 **(1.46)**
 :red_circle: | app/client/src/selectors/onboardingSelectors.tsx | 42.42 **(-40.44)** | 1.43 **(-15.24)** | 21.57 **(-45.1)** | 36.62 **(-41.16)**
 :green_circle: | app/client/src/utils/storage.ts | 34.48 **(0.62)** | 11.11 **(0)** | 7.69 **(0.55)** | 27.56 **(0.75)**
 :x: | ~~app/client/src/components/designSystems/appsmith/CloseButton.tsx~~ | ~~55.56~~ | ~~100~~ | ~~0~~ | ~~55.56~~
 :x: | ~~app/client/src/components/editorComponents/Onboarding/Boxed.tsx~~ | ~~75~~ | ~~28.57~~ | ~~100~~ | ~~75~~
 :x: | ~~app/client/src/components/editorComponents/Onboarding/EndTour.tsx~~ | ~~58.82~~ | ~~50~~ | ~~0~~ | ~~58.82~~
 :x: | ~~app/client/src/components/editorComponents/Onboarding/Helper.tsx~~ | ~~39.77~~ | ~~33.63~~ | ~~0~~ | ~~43.21~~
 :x: | ~~app/client/src/components/editorComponents/Onboarding/Indicator.tsx~~ | ~~55.56~~ | ~~25~~ | ~~22.22~~ | ~~55.56~~
 :x: | ~~app/client/src/components/editorComponents/Onboarding/utils.tsx~~ | ~~57.14~~ | ~~100~~ | ~~50~~ | ~~58.33~~
 :x: | ~~app/client/src/constants/OnboardingConstants.tsx~~ | ~~56.25~~ | ~~50~~ | ~~8.33~~ | ~~56.25~~
 :x: | ~~app/client/src/pages/Editor/Explorer/Onboarding/DBQueryGroup.tsx~~ | ~~47.37~~ | ~~37.5~~ | ~~0~~ | ~~48.65~~
 :x: | ~~app/client/src/pages/Editor/Explorer/Onboarding/Loading.tsx~~ | ~~71.43~~ | ~~100~~ | ~~0~~ | ~~71.43~~
 :x: | ~~app/client/src/pages/Editor/Explorer/Onboarding/index.tsx~~ | ~~48.28~~ | ~~50~~ | ~~0~~ | ~~48.28~~</details>

### CODE REVIEW

File: app/client/src/components/editorComponents/ActionCreator/Fields.tsx [hunk 1]
```diff
@@ -831,7 +831,11 @@ function renderField(props: {
       break;
   }
 
-  return <div key={fieldType}>{viewElement}</div>;
+  return (
+    <div data-guided-tour-iid={field.label} key={fieldType}>
```
Comment (arunvjn) [inline app/client/src/components/editorComponents/ActionCreator/Fields.tsx:835]: Possible type here ("data-guided-tour-id")?

Comment (akash-codemonk) [reply app/client/src/components/editorComponents/ActionCreator/Fields.tsx:835]: iid(indicator id) 😅

File: app/client/src/pages/Editor/GuidedTour/useComputeCurrentStep.ts [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        showIndicator(`[data-guided-tour-iid='onSuccess']`, "left", {
+          top: 0,
+          left: -10,
+        });
+      }
+    }
+  }, [step, hadReachedStep, buttonWidgetSuccessBinding]);
+
+  useEffect(() => {
+    if (step === 9) {
```
Comment (arunvjn) [inline app/client/src/pages/Editor/GuidedTour/useComputeCurrentStep.ts:286]: Can we use an ENUM for the steps with meaningful names ? Can help readability

File: app/client/src/pages/Editor/GuidedTour/utils.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    offset: {
+      top: number;
+      left: number;
+    },
+  ) {
+    if (!primaryReference) {
+      this.destroy();
+      return;
+    }
+    const coords = getCoords(primaryReference);
```
Comment (arunvjn) [inline app/client/src/pages/Editor/GuidedTour/utils.ts:32]: _getCoordinates_ Can we use full name here ?

File: app/client/src/pages/Editor/GuidedTour/utils.ts [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+
+    // Remove previous indicator if it is unable to find the
+    // correct position
+    if (coords.width === 0) {
+      this.destroy();
+      return;
+    }
+
+    if (position === "top") {
```
Comment (arunvjn) [inline app/client/src/pages/Editor/GuidedTour/utils.ts:41]: A comment or a constant to describe what **_18_** is, in the calculations, would help.

File: app/client/src/pages/Editor/Explorer/Actions/MoreActionsMenu.tsx [hunk 5]
```diff
@@ -90,8 +93,14 @@ export function MoreActionsMenu(props: EntityContextMenuProps) {
...
-      dispatch(deleteAction({ id: actionId, name: actionName })),
+    (actionId: string, actionName: string) => {
+      if (guidedTourEnabled) {
+        dispatch(toggleShowDeviationDialog(true));
+        return;
+      }
+
+      dispatch(deleteAction({ id: actionId, name: actionName }));
+    },
     [dispatch],
```
Comment (arunvjn) [inline app/client/src/pages/Editor/Explorer/Actions/MoreActionsMenu.tsx:104]: `guidedTourEnabled` should be in useCallback deps ?

File: app/client/src/pages/Editor/FirstTimeUserOnboarding/IntroductionModal.tsx [hunk 6]
```diff
@@ -27,62 +31,86 @@ const Wrapper = styled.div`
...
+const ModalSubHeader = styled.h5`
   font-size: 20px;
-  margin: 16px 0 6px;
-  text-align: center;
+  margin-top: 20px;
 `;
 
 const ModalBody = styled.div`
-  text-align: center;
+  margin: 21px 54px 16px;
```
Comment (arunvjn) [inline app/client/src/pages/Editor/FirstTimeUserOnboarding/IntroductionModal.tsx:53]: Spaces needs to be in multiples of 4. Can we do this everywhere else in these changes ?

File: app/client/src/utils/AppsmithUtils.tsx [hunk 7]
```diff
@@ -194,11 +194,15 @@ export const stopEventPropagation = (e: any) => {
 export const createNewQueryName = (
   queries: ActionDataState,
   pageId: string,
+  options?: Record<string, string>,
```
Comment (arunvjn) [inline app/client/src/utils/AppsmithUtils.tsx:197]: I don't see options used elsewhere. Can we simply have prefix as a parameter ?

Comment (akash-codemonk) [reply app/client/src/utils/AppsmithUtils.tsx]: Yeah forgot to remove it. It isn't used anywhere.

**Review by arunvjn — COMMENTED** [review]
I have done one round of review and I see that you have resolved all the comments. The rest looks good to me. However, I would recommend getting another review from someone in UI builders pod.

File: app/client/src/pages/Editor/GuidedTour/constants.tsx [hunk 8]
```diff
@@ -0,0 +1,562 @@
...
+            },
+            {
+              type: "TEXT_WIDGET",
+              position: {
+                top: 14,
+                left: 19,
+              },
+              size: {
+                rows: 1 * GRID_DENSITY_MIGRATION_V1,
+                cols: 2.5 * GRID_DENSITY_MIGRATION_V1,
```
Comment (sbalaji1192) [inline app/client/src/pages/Editor/GuidedTour/constants.tsx:187]: Lets not use decimal values, use only whole numbers

File: app/client/src/pages/Editor/GuidedTour/constants.tsx [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+                text: "Country",
+                textAlign: "LEFT",
+                fontStyle: "BOLD",
+                textColor: "#231F20",
+                version: 1,
+                fontSize: "PARAGRAPH",
+              },
+            },
+            {
+              type: "INPUT_WIDGET",
```
Comment (sbalaji1192) [inline app/client/src/pages/Editor/GuidedTour/constants.tsx:199]: we have deprecated the INPUT_WIDGET. We have a new version now, use `INPUT_WIDGET_V2`

File: app/client/src/sagas/WidgetAdditionSagas.ts [hunk 10]
```diff
@@ -229,6 +229,7 @@ export function* getUpdateDslAfterCreatingChild(
     stateParent,
     addChildPayload,
     widgets,
+    addChildPayload.props?.blueprint,
```
Comment (sbalaji1192) [inline app/client/src/sagas/WidgetAdditionSagas.ts:232]: Could you add comment here, that blueprint is only used in onboardingFlow?

File: app/client/src/reducers/uiReducers/onBoardingReducer.ts [hunk 11]
```diff
@@ -1,219 +1,204 @@
...
     state: OnboardingState,
     action: ReduxAction<boolean>,
   ) => {
     return {
       ...state,
-      enableFirstTimeUserOnboarding: action.payload,
+      tableWidgetWasSelected: action.payload,
     };
   },
-  [ReduxActionTypes.SET_FIRST_TIME_USER_ONBOARDING_APPLICATION_ID]: (
```
[the remark this answers is not in the document]
Comment (akash-codemonk) [reply app/client/src/reducers/uiReducers/onBoardingReducer.ts:185]: Yes it is being replaced with the new one in this PR.

**Review by sbalaji1192 — APPROVED** [review]
LGTM

**Pull request merged**